support subscribe

sbawaska opened the pull request:
Adding support for subscribe. This method takes a function as a
parameter which is executed for each message.
closes #2

Commit:
support subscribe

Adding support for subscribe. This method takes a function as a
parameter which is executed for each message.
closes #2

ericbottard commented on client.go:
```diff
@@ -110,6 +110,61 @@ func chopContentType(contentType string) string {
 	return strings.Split(contentType, ";")[0]
 }
 
+func (lc *StreamClient) Subscribe(ctx context.Context, group string, f func(ctx context.Context, payload []byte, contentType string) error) error {
```
Should create a type for the callback fn

sbawaska replied:
done!

ericbottard commented on client.go:
```diff
@@ -110,6 +110,61 @@ func chopContentType(contentType string) string {
+	if err != nil {
+		return err
+	}
+	for true {
```
I think this should happen in a separate goroutine

sbawaska replied:
done!

ericbottard commented on client.go:
```diff
@@ -110,6 +110,61 @@ func chopContentType(contentType string) string {
+		Assignment:           subscribeReply.GetAssignment(),
+		LastKnownOffset:      0,
+	}
+	receiveClient, err := lc.client.Receive(context.Background(), &receiveRequest)
```
Should use ctx?

sbawaska replied:
done!

ericbottard commented on client.go:
```diff
@@ -110,6 +110,61 @@ func chopContentType(contentType string) string {
+		if err != nil {
+			return err
+		}
+		err = f(ctx, m.GetPayload(), m.ContentType)
```
Maybe f should have access to the other headers as well

ericbottard reviewed (changes requested):
Looks good overall.

General comments on the form of the API. I'd really like to see a real word use case of this to decide how it "feels". What about a simple "log" sink?

Also, there should be a way to unsubscribe, maybe via the (cancellable) context which is not really used to its full potential here I think.

sbawaska commented:
> I'd really like to see a real word use case of this to decide how it "feels". What about a simple "log" sink?

Here is an "integration" test I ran to verify the correctness:
```

func TestStreamClient_Publish(t *testing.T) {
	c, err := client.NewStreamClient("localhost:6565", "test1_in", "text/plain")
	if err != nil {
		t.Error(err)
	}
	headers := make(map[string]string)
	reader := strings.NewReader("BAR")
	publishResult, err := c.Publish(context.Background(), reader, nil, "text/plain", headers)
	if err != nil {
		t.Error(err)
	}
	fmt.Printf("Published: %+v", publishResult)
}

func TestStreamClient_Subscribe(t *testing.T) {
	c, err := client.NewStreamClient("localhost:6565", "test1_in", "text/plain")
	if err != nil {
		t.Error(err)
	}
	err = c.Subscribe(context.Background(), "g1", func(ctx context.Context, payload []byte, contentType string) error {
		fmt.Printf("PAYLOAD: %s, contentType: %s\n", string(payload), contentType)
		return nil
	})
	if err != nil {
		t.Error(err)
	}
}
```

Will incorporate the feedback and make this test run in CI today.

Commit:
handle multiple subscriptions

Subscribe now returns a context.CancelFunc that can be called to cancel
the subscription.
Each subscriptionnow runs in its own goroutine.
Enhance error handling

Commit:
adding integration test

markfisher commented on client.go:
```diff
@@ -50,6 +50,9 @@ type PublishResult struct {
 	Offset    uint64
 }
 
+type Subscriber = func(ctx context.Context, payload []byte, contentType string) error
```
maybe this type should be called `EventHandler` or simply `Handler`?

I think of "subscriber" as the one who calls the `Subscribe` function, just as "publisher" is the one who calls the `Publish` function. It just happens that unlike the publisher who's able to send an event with each call, the subscriber is making a single call to register a callback function for handling events.

sbawaska replied:
sounds good. renamed to EventHandler.

Commit:
support cancelling subscription and renamed function to EventHandler

Commit:
remove Println

Commit:
fix label to gateway

ericbottard commented on client.go:
```diff
@@ -50,6 +51,14 @@ type PublishResult struct {
 
+// EventHandler is a function to process the messages read from the stream and is passed as
+// a parameter to the subscribe call.
+type EventHandler = func(ctx context.Context, payload []byte, contentType string) error
```
Need to expose all headers here

sbawaska replied:
fixed!

ericbottard commented on client.go:
```diff
@@ -110,6 +119,80 @@ func chopContentType(contentType string) string {
 	return strings.Split(contentType, ";")[0]
 }
 
+// Subscribe function subscribes for events form the StreamClient TopicName after the given offset. An offset of zero should be
```
This sentence reads weird.

sbawaska replied:
I revised it a bit. Please feel free to suggest an alternative if it still seems off.

ericbottard commented on client.go:
```diff
@@ -110,6 +119,80 @@ func chopContentType(contentType string) string {
+// To deal with errors while reading messages, an error handler function should also be provided.
+//
+// The function returns a context.CancelFunc which may be called for cancelling the subscription.
+func (lc *StreamClient) Subscribe(ctx context.Context, group string, offset uint64, f EventHandler, e EventErrHandler) (context.CancelFunc, error) {
```
Should we group the two handlers in an interface?

ericbottard commented on client.go:
```diff
@@ -110,6 +119,80 @@ func chopContentType(contentType string) string {
+	}
+	subscribedClient, err := lc.client.Subscribe(subContext, &request)
+	if err != nil {
+		return cancel, err
```
The cancelFunc returned should be nil in case of error

sbawaska replied:
That is how it was, changed because `go vet` complained.

ericbottard replied:
Curious to see what `go vet` has to say in detail for this particular case. The general rule is to return zero values in case of errors.

sbawaska replied:
This is what go vet said:
```
./client.go:129:2: the cancel function is not used on all paths (possible context leak)
./client.go:137:3: this return statement may be reached without using the cancel var defined on line 129
```

ericbottard replied:
OK, I think then that we should rather make sure that this function cancels the context itself, not return the cancelFunc (which the user may happily forget to use)

sbawaska replied:
We return cancelFunc so that users can call it to cancel the "subscription". 
It is ok if the users do not call cancelFunc, since all we are doing is wrapping the user passed context. So, when the user passed context is terminated, the context we created will terminate too.

ericbottard commented on client.go:
```diff
@@ -110,6 +119,80 @@ func chopContentType(contentType string) string {
+		for {
+			select {
+			case <- subContext.Done():
+				e(cancel, errors.New("context cancelled"))
```
Is it always because the context was canceled that Done() is closed? Could be eg a timeout on the parent context, no?

sbawaska replied:
fair point. How about `context terminated`?

Commit:
adding headers to EventHandler

Commit:
fixed error message

Commit:
fixed wording

markfisher commented on client.go:
```diff
@@ -110,6 +120,80 @@ func chopContentType(contentType string) string {
 	return strings.Split(contentType, ";")[0]
 }
 
+// Subscribe function should be used to listen for events form the StreamClient TopicName after the given offset. An offset of zero should be
```
```suggestion
// Subscribe function should be used to listen for events from the StreamClient TopicName after the given offset. An offset of zero should be
```

Commit:
Update go doc

Co-Authored-By: Mark Fisher <markfisher@users.noreply.github.com>

ericbottard commented on .github/workflows/ci.yml:
```diff
@@ -32,4 +37,35 @@ jobs:
-      
+
+  integration:
+    name: Integration testing
```
I haven't read in detail what those scripts do, but it seems out-of-proportion to setup k8s, riff, etc just to test a glorified Kafka consumer. Setting up liiklus on some backend (even in-memory) should be sufficient IMO.

sbawaska replied:
> but it seems out-of-proportion to setup k8s, riff, etc just to test a glorified Kafka consumer

Actually, fats make it trivially easy to setup, and given that this is an integration test, I think we should exercise all of the moving parts.

We still need to invest in unit testing this.

ericbottard commented on client.go:
```diff
@@ -50,6 +51,14 @@ type PublishResult struct {
 
+// EventHandler is a function to process the messages read from the stream and is passed as
+// a parameter to the subscribe call.
+type EventHandler = func(ctx context.Context, payload []byte, contentType string, headers map[string]string) error
```
Consider changing payload type to Reader which *could* allow us to stream the payload in the future

sbawaska replied:
done!

ericbottard commented on client.go:
```diff
@@ -50,6 +51,14 @@ type PublishResult struct {
+
+// EventErrHandler is a function to handle errors while reading subscription messages and
+// is passed as a parameter to the subscribe call.
+type EventErrHandler = func(cancel context.CancelFunc, err error)
```
Why does the error handler get the cancelFunc? There are cases below when the error handler gets triggered **because** the context got (already) canceled.

sbawaska replied:
The error handler can choose to cancel the subscription by calling the cancelFunc if it chooses to. I will update the doc.

It is ok to call the cancelFunc multiple times since it then becomes a no-op.

Commit:
update errHandler doc

Commit:
EventHandler now gets io.Reader rather than []byte

Commit:
gather diagnostics

Commit:
increase timeout to allow the gateway get to ready state

ericbottard commented:
Merging as is. Should open another PR to simplify integration testing, simply using a liiklus instance (eg through docker) as target
